src/Simulation/TargetDefinitions/Decompositions/CCNOTFromCCZ.qs

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Microsoft.Quantum.Intrinsic {
+    open Microsoft.Quantum.Diagnostics;
+
+    /// # Summary
+    /// Applies the doubly controlled–NOT (CCNOT) gate to three qubits.
+    ///
+    /// # Input
+    /// ## control1
+    /// First control qubit for the CCNOT gate.
+    /// ## control2
+    /// Second control qubit for the CCNOT gate.
+    /// ## target
+    /// Target qubit for the CCNOT gate.
+    ///
+    /// # Remarks
+    /// Equivalent to:
+    /// ```qsharp
+    /// Controlled X([control1, control2], target);
+    /// ```
+    @EnableTestingViaName("Test.TargetDefinitions.CCNOT")
+    operation CCNOT (control1 : Qubit, control2 : Qubit, target : Qubit) : Unit is Adj + Ctl {
+        body (...) {
+            // [Page 15 of arXiv:1206.0758v3](https://arxiv.org/pdf/1206.0758v3.pdf#page=15)
+            within {
+                H(target);
+            }
+            apply {
+                Controlled Z([control1, control2], target);
+            }
+        }
+        controlled (ctls, ...) {
+            Controlled X (ctls + [control1, control2], target);
```

Comment:
Why does the controlled specialization not use the same decomposition?

Reply:
That's certainly a possibility, and should end up being correct too, but I think the intent of the original author of these decompositions was to connect multi-controlled `CCNOT` back to `X` explicitly so that it was easier to follow the flow of the decomposition. That way much the same way that `Controlled X` with 2 control qubits is exactly `CCNOT` it is obvious that `CCNOT` with n control qubits is identical to `Controlled X` with n + 2 control qubits.

There's certainly opportunity in the future for experimentation and refinement of these decompositions, either in-place or as new target packages altogether, but for now I'd like to get this version in so we can continue to iterate on top of it.